### ISSUE TITLE
StackOverflowError or NoClassDefFoundError is occurred when received form-data with many fields from multipart/form-data

### ISSUE THREAD

**Issue by yousuketto** [playframework/playframework#10590]
### Play Version
2.8.5

### API
Scala

### Operating System
Windows 10 Pro

### JDK
openjdk version "1.8.0_252"
OpenJDK Runtime Environment Corretto-8.252.09.2 (build 1.8.0_252-b09)
OpenJDK 64-Bit Server VM Corretto-8.252.09.2 (build 25.252-b09, mixed mode)

### Library Dependencies
sbt 1.3.13

### Expected Behavior
1. Post form-data with 4,000+ fields from multipart/form-data form.
2. Can execute target action.

### Actual Behavior
StackOverflowError or NoClassDefFoundError is occurred.
```
2020-12-10 06:11:15 ERROR akka.actor.ActorSystemImpl akka.actor.ActorSystemImpl(application) Uncaught error from thread [application-akka.actor.default-dispatcher-10]: null, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[application]
java.lang.StackOverflowError: null
        at java.util.regex.Pattern$BmpCharProperty.match(Pattern.java:3812)
        at java.util.regex.Pattern$Ques.match(Pattern.java:4195)
        at java.util.regex.Pattern$BmpCharProperty.match(Pattern.java:3812)
        at java.util.regex.Pattern$GroupTail.match(Pattern.java:4731)
        at java.util.regex.Pattern$Curly.match0(Pattern.java:4286)
        at java.util.regex.Pattern$Curly.match(Pattern.java:4248)
        at java.util.regex.Pattern$GroupHead.match(Pattern.java:4672)
        at java.util.regex.Pattern$Begin.match(Pattern.java:3539)
        at java.util.regex.Matcher.match(Matcher.java:1270)
        at java.util.regex.Matcher.matches(Matcher.java:604)
        at scala.util.matching.Regex.runMatcher(Regex.scala:342)
        at scala.util.matching.Regex.unapplySeq(Regex.scala:284)
        at play.core.parsers.Multipart$FileInfoMatcher$.$anonfun$unapply$3(Multipart.scala:239)
        at scala.collection.Iterator$$anon$9.next(Iterator.scala:575)
        at scala.collection.mutable.Growable.addAll(Growable.scala:62)
        at scala.collection.mutable.Growable.addAll$(Growable.scala:59)
        at scala.collection.immutable.MapBuilderImpl.addAll(Map.scala:692)
        at scala.collection.immutable.Map$.from(Map.scala:643)
        at scala.collection.IterableOnceOps.toMap(IterableOnce.scala:1263)
        at scala.collection.IterableOnceOps.toMap$(IterableOnce.scala:1262)
        at scala.collection.AbstractIterator.toMap(Iterator.scala:1279)
        at play.core.parsers.Multipart$FileInfoMatcher$.$anonfun$unapply$1(Multipart.scala:245)
        at scala.Option.map(Option.scala:242)
        at play.core.parsers.Multipart$FileInfoMatcher$.unapply(Multipart.scala:245)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.parseHeader(Multipart.scala:435)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.handleDataPart(Multipart.scala:544)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.parseHeader(Multipart.scala:458)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.handleDataPart(Multipart.scala:544)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.parseHeader(Multipart.scala:458)
```
```
2020-12-10 06:07:32 ERROR akka.actor.ActorSystemImpl akka.actor.ActorSystemImpl(application) Uncaught error from thread [application-akka.actor.default-dispatcher-6]: Could not initialize class play.core.parsers.Multipart$NotEnoughDataException$, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[application]
java.lang.NoClassDefFoundError: Could not initialize class play.core.parsers.Multipart$NotEnoughDataException$
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.tryParseInitialBoundary(Multipart.scala:387)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.$anonfun$state$1(Multipart.scala:349)
        at play.core.parsers.Multipart$BodyPartParser$$anon$1.onPush(Multipart.scala:354)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:541)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:423)
        at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:625)
        at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:502)
        at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:600)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:769)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:784)
        at akka.actor.Actor.aroundReceive(Actor.scala:537)
        at akka.actor.Actor.aroundReceive$(Actor.scala:535)
```

### Reproducible Test Case
The error is occurred when `testForPost` action is received form-data with many fields from multipart/form-data form.
```
package controllers

import javax.inject._
import play.api.mvc._

@Singleton
class HomeController @Inject()(val controllerComponents: ControllerComponents) extends BaseController {

  def index() = Action { implicit request: Request[AnyContent] =>
    val checkboxCount = request.getQueryString("checkboxCount").map(_.toInt).getOrElse(1000)
    Ok(views.html.index(checkboxCount))
  }

  def testForPost() = Action(parse.multipartFormData) { implicit request: Request[MultipartFormData[play.api.libs.Files.TemporaryFile]] =>
    Redirect(controllers.routes.HomeController.index)
  }
}
```

I used the following form.
```
@(checkboxCount: Int)(implicit request: RequestHeader)
<!DOCTYPE html>
<html lang="en">
  <body>
    <h1>Welcome to Play!</h1>
    <form action="@controllers.routes.HomeController.testForPost" method="POST" enctype="multipart/form-data" >
      @helper.CSRF.formField
      <button type="submit">submit</button>
      @for(i <- (0 until checkboxCount)) {
      <input type="checkbox" value=true checked name="val@i">
      }
    </form>
  </body>
</html>
```

**Comment by ignasi35** [comment]
There is a trade-off between stack size and payload size. You can workaround this issue increasing the stack size and reducing the `play.http.parser.maxMemoryBuffer` config value.

I could only reproduce the issue on my local machine by forcing a smaller stack size than the default value and increasing the memory buffer. Using default values and trying out your code with a 10_000 checkboxes I would trigger Play's memoryBuffer limits validation: `For request 'POST /testPost' [Memory buffer full on part val1023]`

**Comment by yousuketto** [comment]
Thanks for reply. @ignasi35 

`maxMemoryBuffer` is related value size not only name size.
So, I need some margin if the form has text field.

If there is an attacker that want to down my application, I need to estimate stack size to guard.
I think it is hard work for framework users (at least for me). 
So I want an implementation not to need to estimate stack size.

If the implementation will be not realized, I want hints to estimate stack size.
Do you have hints?

**Comment by ignasi35** [comment]
@yousuketto there's no magic trick, unfortunately. 

As a framework user, one can only know the framework limitations (and related trade-offs) and use cases of the application they are implementing to fine-tune the framework limitations to better serve the business with limited or no risk.

As a framework writer, one can only avoid algorithms that are limited by the stack size. We can try to look into this aspect.

**Comment by mkurz** [comment]
Another report: https://discuss.lightbend.com/t/unfathomable-notenoughdataexception/7556

**Comment by mkurz** [comment]
Fixed in #11360
